CatalaLang/catala
# Clearer errors

KEYCOMMITs

Short description of commit 10872b7cc0a664584992b2bc8e405a557cecf216 "Match the compiler program to the new type of error message content"

- Clean up the "suggestions.ml" utility : implementation of autosuggestions functions. All piece of code that needs suggestions uses suggestions.ml. The whole process is documented with comments.
- Emit modification : if a constructor is not found, the program offers suggestions that adapt with the new error message content type while offering pretty-printing : (modulable suggestions, coloured constructors, ...)
- Clean up compiler/tests & documentation 

Short description of commit 1aa2646f51170cdca354ad3055507ab02aa5a9ce "Improve suggestions"
- Improved implementation of suggestions (suggestion_minimum_levenshtein_distance_association > suggestions.ml)
- Magnifying glass emoji added for suggestions
- Improve display by...
	- Fixing "Pos.ml" (pretty-printing box not closed)
	- Remove unnecessary spaces
- compiler/surface/parserdriver.ml
	- Optionnal suggestions added when raising a parser error
	- Function "levenshtein_distance" removed and replaced by "Suggestions.levenshtein_distance"
- Tests updated (mostly line breaks and spacing changes)
- Variable names renamed for greater clarity
- (Based on commit https://github.com/Aminata-Dev/catala/commit/0813f5c36e2ce082314fa4dda85cb0ae8a04bcf6)

Short description of commit affdc104662588ab2ede1fb5d7186e02a7b6b5f5 "Improve the implementation of suggestions"
- suggestions.ml : improve details + algorithm implementation (see @Altgr reviews) + comments
- suggestions implementation : since we want to keep the same structure to display suggestion messages, Suggestion is no longer a formatted string but a string list. The goal is to have a similar formatted string for each suggestion, no matter what the error message is.
--> This involves changes to suggestions.mli, compiler/desugared/from_surface.ml (most important one) and compiler/surface/parser_driver.ml

Short description of commit e47e21b21129a0837ebcb039da98a503286a1d13 "Better help for the user"
- Clearer (more detailed) error messages (see parser.messages, parser_driver.ml ->)
	--> Parser_driver.ml : we're more precise and show all possible instructions to the user.
- Suggestions.ml : unimportant display changes + renaming

Short description of commit 94f8eac85886f85fdffdd55c9da304ebcd93b1c8 "Better indications for the user"
- We're able to say from the parser what the user could have written. It may not be complete due to the acceptable function of Menhir : it is only an indication given to the user (and not intended to be an adaptive documentation)
- .mli file added : module interface for suggestions
- Add a test that provides a typographical or a logical error.
- Documentation and type / name changes for suggestions
